FIO Job File support

saibulusu opened the pull request:
Adding Job File support to the FIOExecutor class. Created a profile PERF-IO-FIO-OLTP to expose this feature. Tested on linux-x64, linux-arm64, windows-x64 VMs.

yangpanMS commented:
Could we include unit test for the fio job file parsing?

yangpanMS commented on website/docs/guides/0011-profiles.md:
Please add document specific to FIO

saibulusu replied:
Added a profile section for PERF-IO-FIO-OLTP.json.

yangpanMS commented on src/VirtualClient/VirtualClient.Actions/FIO/FioDiscoveryExecutor.cs:
```diff
@@ -52,6 +52,23 @@ public List<string> BlockSize
         }
 
+        /// <summary>
+        /// The size of the test file that should use in workload tests (e.g. 496GB).
```
Is this used anywhere?

saibulusu replied:
FileSize was inherited from DiskWorkloadExecutor (via FioExecutor) and used within FioDiscoveryExecutor. I removed the dependency between FioExecutor and DiskWorkloadExecutor, so I had to add this field here.
It is used in the ExecuteAsync method in DiskWorkloadExecutor.

saibulusu commented:
> Could we include unit test for the fio job file parsing?

I have added some unit tests now. None of the unit tests referenced the architecture except for the one I added, so I set the architecture to X64.

saibulusu commented:
The FioExecutor currently appends --ioengine & --filename to the command for the command line approach. I think it would make sense to keep the same in the job file approach, so I think we should expect the template job file that is provided to exclude the ioengine & file path parameters.

saibulusu commented on website/docs/workloads/fio/fio-profiles.md:
```diff
@@ -426,4 +426,89 @@ This profile uses an algorithm to determine the amount of IOPS to run against th
+Sequential IO : It represents the logs of OLTP-C workload.
+Therefore, they are performed on different disks
+
+The intent of this profile is to expose the ability to use template job files via the FioExecutor.
```
Removed this sentence.

saibulusu commented on website/docs/guides/0011-profiles.md:
```diff
@@ -68,6 +68,7 @@ the system. The following table describes the set of well-known parameters that
 | SystemMemoryKilobytes                 | Represents the total memory/RAM (in kilobytes) on the system. Note that industry standard memory unit definitions are used (e.g. 1 kilobyte = 1024 bytes). |
 | SystemMemoryMegabytes                 | Represents the total memory/RAM (in megabytes) on the system. Note that industry standard memory unit definitions are used (e.g. 1 megabyte = 1024 kilobytes or 1024 x 1024 bytes). |
 | SystemMemoryGigabytes                 | Represents the total memory/RAM (in gigabytes) on the system. Note that industry standard memory unit definitions are used (e.g. 1 gigabyte = 1024 megabytes or 1024 x 1024 kilobytes or 1024 x 1024 x 1024 bytes). |
+| ScriptPath:\{script_folder\}			| Represents the path to a script that is copied to the build output 'scripts' directory. |
```
Changed to alphabetical order.

saibulusu commented on src/VirtualClient/VirtualClient.Main/profiles/PERF-IO-FIO-OLTP.json:
```diff
@@ -0,0 +1,282 @@
+        "JobFiles": "{ScriptPath:fio}/oltp-c2.fio.jobfile",
+        "FileName": "fio-test.dat",
+        "DeleteTestFilesOnFinish": false,
+        "DiskFilter": "$.Parameters.DiskFilter",
```
Removed these parameters, some are directly set in the template job file and others are not needed at all.

saibulusu commented on src/VirtualClient/VirtualClient.Main/profiles/PERF-IO-FIO-OLTP.json:
```diff
@@ -0,0 +1,282 @@
+  },
+  "Parameters": {
+    "DiskFillSize": "10G",
+    "FileSize": "10G",
```
I set it to 124 GB and 20 GB to match the multithroughput profile, which runs OLTP as well and has a very similar template job file.

saibulusu commented on website/docs/workloads/fio/fio-profiles.md:
```diff
@@ -426,4 +426,89 @@ This profile uses an algorithm to determine the amount of IOPS to run against th
+## PERF-IO-FIO-OLTP.json
+Runs an IO-intensive workload using the Flexible IO Tester (FIO) toolset. Multi-throughput OLTP-C workload to emulate a SQL Server OLTP disk 
+workload by running four workload compononents in-parallel: random reads, random writes, sequential reads and sequential writes each with an overall 
+weight/percentage defined. 
```
Removed the word "defined" and showed the example for the values that are used in this profile.

saibulusu commented on website/docs/workloads/fio/fio-profiles.md:
```diff
@@ -426,4 +426,89 @@ This profile uses an algorithm to determine the amount of IOPS to run against th
+
+  | Parameter                 | Purpose                                                                         | 
+  |---------------------------|---------------------------------------------------------------------------------|
+  | DefaultRandomIOBlockSize  | Default Block size value for Random Read and Write. |
```
Removed those parameters.
